arangod/Aql/Functions.cpp

```diff
@@ -1288,6 +1347,232 @@ AqlValue Functions::Substring(arangodb::aql::Query* query,
+    if (Search.isArray()) {
+      for (auto const& it : VPackArrayIterator(slice)) {
+        if (!it.isString()) {
+          RegisterInvalidArgumentWarning(query, "SUBSTITUTE");
+          return AqlValue(AqlValueHintNull());
+        }
+        arangodb::velocypack::ValueLength length;
+        const char *str = it.getString(length);
+        match.push_back(UnicodeString(str, length));
+      }
+    }
+    else {
+      if (!Search.isString()) {
+        RegisterInvalidArgumentWarning(query, "SUBSTITUTE");
+        return AqlValue(AqlValueHintNull());
+      }
+      arangodb::velocypack::ValueLength length;
+      const char *str = slice.getString(length);
+      match.push_back(UnicodeString(str, length));
+    }
+    if (parameters.size() > 2) {
+      AqlValue Replace = ExtractFunctionParameterValue(parameters, 2);
+      VPackSlice rslice = materializer.slice(Replace, false);
+      if (Replace.isArray()) {
+        for (auto const& it : VPackArrayIterator(rslice)) {
+          if (!it.isString()) {
+            RegisterInvalidArgumentWarning(query, "SUBSTITUTE");
+            return AqlValue(AqlValueHintNull());
+          }
+          arangodb::velocypack::ValueLength length;
+          const char *str = it.getString(length);
+          replace.push_back(UnicodeString(str, length));
+        }
+      }
+      else if (Replace.isString()) {
+        // If we have a string as replacement,
+        // it counts in for all found values.
+        replaceWasPlainString = true;
+        arangodb::velocypack::ValueLength length;
+        const char *str = rslice.getString(length);
+        replace.push_back(UnicodeString(str, length));
+      }
+      else {
+        RegisterInvalidArgumentWarning(query, "SUBSTITUTE");
+        return AqlValue(AqlValueHintNull());
+      }
+    }
+  }
+
+  AqlValue value = ExtractFunctionParameterValue(parameters, 0);
+  if ((limit == 0) || (match.size() == 0)) {
+    // if the limit is 0, or we don't have any match pattern, return the source string.
+    return AqlValue(value);
+  }
+
+  transaction::StringBufferLeaser buffer(trx);
+  arangodb::basics::VPackStringBufferAdapter adapter(buffer->stringBuffer());
+
+  AppendAsString(trx, adapter, value);
+  UnicodeString unicodeStr(buffer->c_str(), static_cast<int32_t>(buffer->length()));
+
+  auto locale = LanguageFeature::instance()->getLocale();
+  std::vector<std::unique_ptr<StringSearch>> searchVec;
+  searchVec.reserve(match.size());
+  UErrorCode status = U_ZERO_ERROR;
+  for (auto const& searchStr : match) {
+    // create a vector of string searches
+    searchVec.push_back(std::make_unique<StringSearch>(searchStr, unicodeStr, locale, nullptr, status));
+  }
+  
+  std::vector<std::pair<int32_t, int32_t>> srchResultPtrs;
+  std::string utf8;
+  srchResultPtrs.reserve(match.size());
+  for (auto& search : searchVec) {
+    // We now find the first hit for each search string. 
+    auto pos = search->first(status);
+    int32_t len;
```

Comment:
Should this be initialized for the case that  `POS == USEARCH_DONE`
Will pair that have USEARCH_DONE in `.first` be ignored later on? If so why add them in the first place?

Reply:
the number in the vectors has to be identical, so counters can be easily be used in the loop below.